lib/active_merchant/billing/gateways/place_to_pay.rb

```diff
@@ -0,0 +1,196 @@
+        post[:currency] = (options[:currency] || currency(money))
+      end
+
+      def build_url(action, base)
+        url = base
+        url += action
+        url
+      end
+
+      def commit(method, action, parameters)
+        base_url = (test? ? test_url : live_url)
+        url = build_url(action, base_url)
+
+        begin
+          url = (test? ? test_url : live_url) + action.to_s
+          rel_path = "#{method}/v1/#{action}"
+          response = api_request(method, url, rel_path, parameters)
+
+        rescue ResponseError => e
+          raw_response = e.response.body
+          response = parse(raw_response)
+        end
+
+        Response.new(
+          success_from(action, response, options),
+          message_from(response),
+          response,
+          test: test?,
+          error_code: error_code_from(action, response, options),
+          authorization: response['authorization'],
+          network_transaction_id: response['internalReference'],
+        )
+      end
+
+
+      def success_from(action, response, options)
+        case action
+        when :authonly.to_s, :lookup.to_s, :mpiQuery.to_s, :interests.to_s, :generate_otp.to_s,
+          :validate_otp.to_s, :capture.to_s, :search.to_s
+          return response['status'] && response['status']['status'] === RESPONSES_STATUS[:ok]
+        when :sale.to_s, :query_transaction.to_s, :refund.to_s
+          return response['status'] && response['status']['status'] === RESPONSES_STATUS[:approved]
+        else
+          return response['status']['status'];
+        end
+      end
+
+      def message_from(response)
+        response['status'] && response['status']['message'] ? response['status']['message'] :
+          COMMON_MESSAGES[:missed]
+      end
+
+      def authorization_from(response); end
+
+      def post_data(action, parameters = {})
+        JSON.generate(parameters)
+      end
+
+      def error_code_from(action, response, options)
+        unless success_from(action, response, options)
+          # TODO: lookup error code for this response
+          STANDARD_ERROR_CODE_MAPPING[response['reason']]
+        end
+      end
+
+      def add_auth(post, options)
+        original_nonce = generate_nonce()
+        seed = get_date_in_iso_format()
+        tranKey = generate_trans_key(original_nonce, seed)
+        post[:auth] = {
+          login: @login,
+          tranKey: tranKey,
+          seed: seed,
+          nonce: convert_to_Base64(original_nonce)
+        }
+      end
+
+      def add_payer(post, options)
+        post[:payer] = options[:payer]
+      end
+
+      def add_payment(post, options)
+        post[:payment] = options[:payment]     
+      end
+
+      def add_instrument(post, payment, options)
+        post[:instrument] = options[:instrument]
+        post[:instrument][:card][:number] = payment.number;
+        post[:instrument][:card][:expiration] = (payment.month).to_s + "/" + (payment.month).to_s;
```

Comment:
Should the second `payment.month` be `payment.year` here?

Also, there are several sub-objects within the instrument object in the gateway docs. Shouldn't these be added? If there will be any customers transacting from Europe then the 3DS fields will be required.

Reply:
Fixed! Thanks for the heads up